streams: Send stream creation event when changing stream to public.

sahil839 opened the pull request:
This PR adds code to send stream creation and peer add events
when stream is changed from private to public. These events are
only sent to users who are not susbcribed to the stream and are
not realm admins as subscribers and realm admins already have
the stream data. This will update the stream data with clients
and will remove the need to reload to view the modified stream.


<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.--> #22194

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![stream-privacy-event](https://user-images.githubusercontent.com/35494118/172408983-be09313d-67e9-4188-88e4-57427a9e0d4b.gif)

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented on zerver/actions/streams.py:
```diff
@@ -853,6 +854,27 @@ def do_change_stream_permission(
+        old_can_access_stream_user_ids = stream_subscribers | {
+            user.id for user in stream.realm.get_admin_users_and_bots()
+        }
+        non_guest_user_ids = active_non_guest_user_ids(stream.realm_id)
```
Did not directly use `can_access_stream_user_ids` here to get the users who can access the stream after changing, because it includes guest subscriptions too and will make an extra query for it but we do not need it since we anyway want to exclude subscribed users at the end.

sahil839 commented:
This is implemented like we do when a user is subscribed to private stream, there also we send a stream creation event because user was not having access to stream previously.

alya commented:
@timabbott I'll leave this to you to review, unless you specifically would like me to take a look.

sahil839 commented:
- Added commit to update the stream objects in `unsubscribed` and `never_subscribed` lists as well.
- Updated the original commit to not send stream creation events to users who were previously subscribed to the stream since those users already have the stream object.

@timabbott this is ready for your review.

timabbott commented on zerver/actions/streams.py:
```diff
@@ -853,6 +853,42 @@ def do_change_stream_permission(
+            stream_ids=[stream.id],
+            user_ids=sorted(stream_subscriber_user_ids),
+        )
+        send_event(stream.realm, peer_add_event, peer_notify_user_ids)
```
Should the users in `notify_stream_creation_ids` be excluded from the events emitted later in this function?

sahil839 replied:
Yes, those should be excluded, updated the PR.

sahil839 commented:
@timabbott updated the PR as mentioned above.

timabbott commented:
Merged, thanks for fixing this @sahil839!
